### PR TITLE
feat(core): add storage mechanisms

### PR DESCRIPTION
**Pull request by israx**
#### Description of changes

Adds the following storage mechanisms:

LocalStorage
InMemoryStorage
SessionStorage
CookieStorage
Async Storage

#### Checklist

- [x] PR description included
- [ ] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/core/src/StorageHelper/reactnative.ts [hunk 1]
```diff
@@ -100,3 +101,78 @@ export class StorageHelper {
 		return this.storageWindow;
 	}
 }
+
+class AsyncStorageClass implements KeyValueStorageInterface {
```
Comment (elorzafe) [inline packages/core/src/StorageHelper/reactnative.ts:105]: I would add a `TODO` that this needs to be included on `react-native` Amplify package

File: packages/core/src/StorageHelper/index.ts [hunk 2]
```diff
@@ -99,3 +104,202 @@ export class MemoryKeyValueStorage implements KeyValueStorageInterface {
...
+	 * This is used to clear the storage
+	 * @returns {string} nothing
+	 */
+	async clear(): Promise<void> {
+		if (!this.storage) throw PlatformNotSupportedError;
+		this.storage.clear();
+	}
+}
+
+export const SessionStorage = new SessionStorageClass();
```
Comment (elorzafe) [inline packages/core/src/StorageHelper/index.ts:305]: What if this is an abstract class, the logic is the same except the constructor, what do you think?

File: packages/core/src/StorageHelper/index.ts [hunk 3]
```diff
@@ -1,7 +1,12 @@
 // Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 // SPDX-License-Identifier: Apache-2.0
-
-import { KeyValueStorageInterface } from '../types';
+import * as Cookies from 'js-cookie';
```
Comment (elorzafe) [inline packages/core/src/StorageHelper/index.ts:3]: Would this be tree shakable? maybe we need to add this to a different module?

File: packages/core/package.json [hunk 4]
```diff
@@ -72,7 +72,8 @@
 		"tslib": "^2.5.0",
 		"universal-cookie": "^4.0.4",
 		"zen-observable-ts": "0.8.19",
-		"rxjs": "^7.8.1"
+		"rxjs": "^7.8.1",
+		"js-cookie": "3.0.5"
```
Comment (elorzafe) [inline packages/core/package.json:76]: Do we need to add another dependency? could be possible to not have this?

**Review by elorzafe — CHANGES_REQUESTED** [review]
Biggest concern is adding another dependency and maybe split the helpers in separate files just in case there are issue with tree shaking.

File: packages/core/package.json (hunk 4 above)
Comment (israx) [reply packages/core/package.json:76]: Yeah I think we discussed this in the design already. Pretty much `js-cookie` is to handle the overwhelming cookies operations. The package is pretty light weight and well supported: https://www.npmjs.com/package/js-cookie

File: packages/core/src/StorageHelper/index.ts (hunk 2 above)
Comment (israx) [reply packages/core/src/StorageHelper/index.ts]: I don't think an abstract class will be the right way. It seems to me that it is more like a hacky thing. I mean, yeah abstract classes can't be instantiated but they are mainly used for OOP scenarios. The other solution is just to have a class with static members.

**Review by elorzafe — DISMISSED** [review]
Thanks @israx 

Good work 🏅

File: packages/core/package.json (hunk 4 above)
Comment (jimblanc) [reply packages/core/package.json:76]: Looks good but we might want to loosen up the dependency in-case customers have this in their graph as well

```suggestion
		"js-cookie": "^3.0.5"
```

File: packages/core/src/index.ts [hunk 5]
```diff
@@ -43,7 +43,14 @@ export {
...
 } from './types';
-export { StorageHelper, MemoryStorage } from './StorageHelper';
+export {
+	StorageHelper,
+	MemoryStorage,
+	LocalStorage,
+	CookieStorage,
+	SessionStorage,
+	MemoryKeyValueStorage,
+} from './StorageHelper';
```
Comment (jimblanc) [inline packages/core/src/index.ts:53]: Do we need to export these from the umbrella package as well so customers can customize them?

Here: https://github.com/aws-amplify/amplify-js/blob/next/packages/aws-amplify/src/utils/index.ts (`aws-amplify/utils` route)

**Review by jimblanc — COMMENTED** [review]
Looks good! Some small comments below

**Review by jimblanc — APPROVED** [review]
Looks good!

**Review by elorzafe — APPROVED** [review]

**Pull request merged**